File: src/subbands.py

import numpy as np

from src.config import (
    SAMPLE_RATE,
    FFT_SIZE,
    NUM_FREQ_BINS,
    NUM_SUBBANDS
)


def hz_to_bark(frequency):
    """
    Convert frequency in Hz to Bark scale.
    """

    return (
        13.0 * np.arctan(0.00076 * frequency)
        + 3.5 * np.arctan(
            (frequency / 7500.0) ** 2
        )
    )


def bark_to_hz(bark):
    """
    Convert Bark scale back to frequency in Hz.
    """

    frequencies = np.linspace(
        0,
        SAMPLE_RATE / 2,
        10000
    )

    barks = hz_to_bark(frequencies)

    return np.interp(
        bark,
        barks,
        frequencies
    )


def create_bark_filterbank():
    """
    Create 22 triangular Bark filters.

    Returns:
        filters:
            Shape (22, 257)
    """

    min_bark = hz_to_bark(0)

    max_bark = hz_to_bark(
        SAMPLE_RATE / 2
    )

    bark_points = np.linspace(
        min_bark,
        max_bark,
        NUM_SUBBANDS + 2
    )

    hz_points = bark_to_hz(
        bark_points
    )

    bin_points = np.floor(
        (FFT_SIZE + 1)
        * hz_points
        / SAMPLE_RATE
    ).astype(int)

    bin_points = np.clip(
        bin_points,
        0,
        NUM_FREQ_BINS - 1
    )

    filters = np.zeros(
        (
            NUM_SUBBANDS,
            NUM_FREQ_BINS
        ),
        dtype=np.float32
    )

    for band in range(NUM_SUBBANDS):

        left = bin_points[band]

        center = bin_points[band + 1]

        right = bin_points[band + 2]

        # Rising slope
        if center > left:

            for k in range(
                left,
                center
            ):

                filters[band, k] = (
                    (k - left)
                    / (center - left)
                )

        # Falling slope
        if right > center:

            for k in range(
                center,
                right
            ):

                filters[band, k] = (
                    (right - k)
                    / (right - center)
                )

    return filters


def get_bark_band_centers():
    """
    Return the center frequency of each
    Bark sub-band.

    Returns:
        numpy array with shape (22,)
    """

    min_bark = hz_to_bark(0)

    max_bark = hz_to_bark(
        SAMPLE_RATE / 2
    )

    bark_points = np.linspace(
        min_bark,
        max_bark,
        NUM_SUBBANDS + 2
    )

    hz_points = bark_to_hz(
        bark_points
    )

    # The center points are everything
    # except the first and last boundary.

    center_frequencies = (
        hz_points[1:-1]
    )

    return center_frequencies.astype(
        np.float32
    )


def extract_subband_features(complex_stft):
    """
    Convert complex STFT into 44-dimensional
    Bark + temporal features.

    Input:
        complex_stft:
            Shape (257, time_frames)

    Output:
        features:
            Shape (time_frames, 44)

        First 22:
            log Bark-band energy

        Next 22:
            temporal energy difference

    The first frame has no previous frame,
    so its temporal difference is defined as zero.
    """

    # -----------------------------------------
    # 1. Magnitude
    # -----------------------------------------

    magnitude = np.abs(
        complex_stft
    )

    # -----------------------------------------
    # 2. Power spectrum
    # -----------------------------------------

    power = magnitude ** 2

    # -----------------------------------------
    # 3. Bark filterbank
    # -----------------------------------------

    filters = create_bark_filterbank()

    # -----------------------------------------
    # 4. Calculate energy in each Bark band
    # -----------------------------------------

    band_power = filters @ power

    # -----------------------------------------
    # 5. Convert to logarithmic scale
    # -----------------------------------------

    log_band_power = (
        10.0
        * np.log10(
            band_power + 1e-10
        )
    )

    # -----------------------------------------
    # 6. Calculate temporal change
    # -----------------------------------------
    #
    # We keep the same number of frames
    # as the STFT.
    #
    # Frame 0:
    #     temporal difference = 0
    #
    # Frame 1:
    #     frame 1 - frame 0
    #
    # Frame 2:
    #     frame 2 - frame 1
    #
    # etc.
    # -----------------------------------------

    temporal_difference = np.zeros_like(
        log_band_power
    )

    temporal_difference[:, 1:] = (
        log_band_power[:, 1:]
        - log_band_power[:, :-1]
    )

    # -----------------------------------------
    # 7. Energy features
    # -----------------------------------------
    #
    # Keep ALL frames.
    #
    # Shape:
    #     (22, time_frames)
    # -----------------------------------------

    energy_features = (
        log_band_power
    )

    # -----------------------------------------
    # 8. Combine
    # -----------------------------------------
    #
    # 22 energy features
    #
    # +
    #
    # 22 temporal features
    #
    # =
    #
    # 44 features
    #
    # Shape before transpose:
    #     (44, time_frames)
    # -----------------------------------------

    features = np.concatenate(
        [
            energy_features,
            temporal_difference
        ],
        axis=0
    )

    # -----------------------------------------
    # 9. Transpose
    # -----------------------------------------
    #
    # Before:
    #     (44, time_frames)
    #
    # After:
    #     (time_frames, 44)
    # -----------------------------------------

    features = features.T

    return features.astype(
        np.float32
    )

File: firmware/esp32_impulse_guard/src/subbands.cpp
#include "subbands.h"

#include <math.h>
#include <string.h>

float filters[NUM_SUBBANDS][257];

static float hz_to_bark(float frequency) {
    return (
        13.0f * atanf(0.00076f * frequency)
        + 3.5f * atanf(
            powf(frequency / 7500.0f, 2.0f)
        )
    );
}

static float bark_to_hz(float bark) {
    const int N = 10000;

    float previous_bark = hz_to_bark(0.0f);
    float previous_hz = 0.0f;

    for (int i = 1; i < N; i++) {

        float hz =
            (8000.0f * i) / (N - 1);

        float current_bark =
            hz_to_bark(hz);

        if (current_bark >= bark) {

            float ratio =
                (bark - previous_bark) /
                (current_bark - previous_bark);

            return previous_hz +
                   ratio * (hz - previous_hz);
        }

        previous_bark = current_bark;
        previous_hz = hz;
    }

    return 8000.0f;
}

void subbands_init() {

    memset(
        filters,
        0,
        sizeof(filters)
    );

    const float min_bark =
        hz_to_bark(0.0f);

    const float max_bark =
        hz_to_bark(8000.0f);

    float bark_points[NUM_SUBBANDS + 2];
    int bin_points[NUM_SUBBANDS + 2];

    for (int i = 0;
         i < NUM_SUBBANDS + 2;
         i++) {

        bark_points[i] =
            min_bark +
            (max_bark - min_bark) *
            ((float)i /
             (NUM_SUBBANDS + 1));

        float hz =
            bark_to_hz(
                bark_points[i]
            );

        int bin =
            (int)floorf(
                513.0f * hz / 16000.0f
            );

        if (bin < 0)
            bin = 0;

        if (bin > 256)
            bin = 256;

        bin_points[i] = bin;
    }

    for (int band = 0;
         band < NUM_SUBBANDS;
         band++) {

        int left =
            bin_points[band];

        int center =
            bin_points[band + 1];

        int right =
            bin_points[band + 2];

        if (center > left) {

            for (int k = left;
                 k < center;
                 k++) {

                filters[band][k] =
                    ((float)(k - left)) /
                    ((float)(center - left));
            }
        }

        if (right > center) {

            for (int k = center;
                 k < right;
                 k++) {

                filters[band][k] =
                    ((float)(right - k)) /
                    ((float)(right - center));
            }
        }
    }
}


// ============================================================
// STREAMING FEATURE EXTRACTION
// ============================================================

void extract_subband_features(
    const float* real,
    const float* imag,
    float* features
) {
    static float previous_log_band_power[
        NUM_SUBBANDS
    ];

    static bool first_frame = true;

    float log_band_power[
        NUM_SUBBANDS
    ];


    // --------------------------------------------------------
    // Calculate 22 Bark-band log energies
    // --------------------------------------------------------

    for (int band = 0;
         band < NUM_SUBBANDS;
         band++) {

        float band_power = 0.0f;

        for (int k = 0;
             k < 257;
             k++) {

            float magnitude =
                sqrtf(
                    real[k] * real[k] +
                    imag[k] * imag[k]
                );

            float power =
                magnitude * magnitude;

            band_power +=
                filters[band][k] *
                power;
        }

        log_band_power[band] =
            10.0f *
            log10f(
                band_power + 1e-10f
            );
    }


    // --------------------------------------------------------
    // Features 0..21
    // Current Bark log energies
    // --------------------------------------------------------

    for (int band = 0;
         band < NUM_SUBBANDS;
         band++) {

        features[band] =
            log_band_power[band];
    }


    // --------------------------------------------------------
    // Features 22..43
    // Temporal differences
    // --------------------------------------------------------

    if (first_frame) {

        for (int band = 0;
             band < NUM_SUBBANDS;
             band++) {

            features[
                NUM_SUBBANDS + band
            ] = 0.0f;

            previous_log_band_power[band] =
                log_band_power[band];
        }

        first_frame = false;

    } else {

        for (int band = 0;
             band < NUM_SUBBANDS;
             band++) {

            features[
                NUM_SUBBANDS + band
            ] =
                log_band_power[band]
                -
                previous_log_band_power[band];

            previous_log_band_power[band] =
                log_band_power[band];
        }
    }
}